### ISSUE TITLE
Support for JSX and TSX

### ISSUE THREAD

**Issue by sciencesakura** [tyru/caw.vim#91]
In a file whose `filetype` is `javascript`, Caw.vim uses JavaScript's style comment always.

But if the JavaScript file contains the JSX's style code snippets, then it must be distinguished between the JavaScript code and the JSX one to comment out. The examples are as follows.

Uncommented:

```jsx
const foo = 'this is js style';

const bar = (
    <h1>this is jsx style</h1>
);
```

Commented:

```jsx
/* const foo = 'this is js style'; */

const bar = (
    {/* <h1>this is jsx style</h1> */}
);
```

or

```jsx
// const foo = 'this is js style';

const bar = (
    {
    // <h1>this is jsx style</h1>
    }
);
```

It is the same as the TSX (JSX in TypeScript).

Thanks!

**Comment by tyru** [comment]
Could you try this one?

```vim
autocmd BufNewFile,BufRead *.jsx set ft=javascript.jsx
autocmd BufNewFile,BufRead *.tsx set ft=typescript.tsx
autocmd FileType javascript.jsx call s:set_jsx_comment()
autocmd FileType typescript.tsx call s:set_jsx_comment()

function! s:set_jsx_comment() abort
  let &l:commentstring = ''
  unlet! b:caw_oneline_comment
  let b:caw_wrap_oneline_comment = ['{/*', '*/}']
endfunction
```

The above setting disables oneline comment (`// foo`), so `gcc` just wrap / unwrap current line like `{/* foo */}` for `*.jsx` or `*.tsx`.
If you want to change filename extensions to apply, please edit the first two lines.

**Comment by sciencesakura** [comment]
I tried it. It works to use `{/* foo */}` style comment ALWAYS. But it's wrong in JavaScript code.

**Comment by tyru** [comment]
ah sorry, I misunderstood.
If you want to change comment string conditionally by the cursor position,
I want the information that caw.vim can determine JSX code or JavaScript code, like syntax highlight group.
Have you installed syntax highlight plugin for JSX?

**Comment by sciencesakura** [comment]
I'm using [MaxMEllon/vim-jsx-pretty](https://github.com/MaxMEllon/vim-jsx-pretty).

**Comment by tyru** [comment]
Thanks!

**Comment by davidosomething** [comment]
any update on this?
with the syntax plugin `neoclide/vim-jsx-improve`
I have `jsxTagName jsxTag jsxRegion`

**Comment by tyru** [comment]
@sciencesakura @davidosomething could you try #111 ?

**Comment by sciencesakura** [comment]
@tyru Thanks! But a problem occurred.

my env: macOS 10.15.3 / vim 8.2.0350 installed by homebrew / context_filetype.vim ([latest](https://github.com/Shougo/context_filetype.vim/commit/f200fe69939089da9e61bd9a3ff75b4ef7adc708))

I've tried `<Plug>(caw:wrap:comment)` action when the cursor was on the JavaScript code line — not JSX one — in a `javascript` file. And then the line has been commented using the JSX style `{/* ... */}` unfortunately.

```javascript
// before
const a = 123;

// after
{/* const a = 123; */}

// expected
/* const a = 123; */
```

**Comment by sciencesakura** [comment]
[`s:wrap_oneline.get_comments`](https://github.com/tyru/caw.vim/blob/f63268a5938c6d3ea1c095eeeb79e89e91aea168/autoload/caw/comments/wrap_oneline.vim#L14) returns `[['{/*', '*/}'], ['/*', '*/']]` that contains both JavaScript's style and JSX's one, but only a first element `['{/*', '*/}']` is used always in [`s:wrap.comment_normal`](https://github.com/tyru/caw.vim/blob/f63268a5938c6d3ea1c095eeeb79e89e91aea168/autoload/caw/actions/wrap.vim#L41).

**Comment by tyru** [comment]
@sciencesakura fixed. could you try again?

**Comment by tyru** [comment]
@sciencesakura sorry, merged to `devel` branch. please try this on the branch.

**Comment by sciencesakura** [comment]
I tried it. Once I commented a JSX code,  JavaScript code were commented using JSX style too.

![Untitled 3](https://user-images.githubusercontent.com/27776522/76675586-24b28800-65fe-11ea-9e10-5608dccdc172.gif)

**Comment by tyru** [comment]
Fixed in #118

**Comment by sciencesakura** [comment]
Thanks!! 😆
